### PR TITLE
Gas refund complex calls

### PR DESCRIPTION
**Pull request by miiu96**
Update transaction `fee` and `gasUsed` fields when we find a SCR with refund value

### CODE REVIEW

File: data/transaction.go [hunk 1]
```diff
@@ -114,4 +114,17 @@ type PreparedResults struct {
 	AlteredAccts AlteredAccountsHandler
 	Tokens       []*TokenInfo
 	TxHashStatus map[string]string
+	TxHashRefund map[string]string
+}
+
+// ResponseTransactions is the structure for the transactions response
+type ResponseTransactions struct {
+	Docs []ResponseTransactionDB `json:"docs"`
```
Comment (iulianpascalau) [inline data/transaction.go:122]: use a pointer here? Like `Docs []*ResponseTransactionDB` (json unmarshaling should work)

File: mock/dbTransactionsHandlerStub.go [hunk 2]
```diff
@@ -16,6 +16,10 @@ type DBTransactionProcessorStub struct {
 	SerializeScResultsCalled             func(scrs []*data.ScResult) ([]*bytes.Buffer, error)
 }
 
+func (tps *DBTransactionProcessorStub) SerializeTransactionWithRefund(_ map[string]*data.Transaction, _ map[string]string) ([]*bytes.Buffer, error) {
```
Comment (iulianpascalau) [inline mock/dbTransactionsHandlerStub.go:19]: missing comment

File: process/elasticProcessor.go [hunk 3]
```diff
@@ -441,6 +446,39 @@ func (ei *elasticProcessor) SaveTransactions(
...
+	}
+
+	responseTransactions := &data.ResponseTransactions{}
+	err := ei.elasticClient.DoMultiGet(txsHashes, elasticIndexer.TransactionsIndex, true, responseTransactions)
+	if err != nil {
+		return err
+	}
+
+	txsFromDB := make(map[string]*data.Transaction)
+	for idx := 0; idx < len(responseTransactions.Docs); idx++ {
```
Comment (iulianpascalau) [inline process/elasticProcessor.go:465]: :+1: for iterating over a value slice.

File: data/transaction.go (hunk 1 above)
Comment (miiu96) [reply data/transaction.go:122]: done.

File: mock/dbTransactionsHandlerStub.go (hunk 2 above)
Comment (miiu96) [reply mock/dbTransactionsHandlerStub.go:19]: added

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

**Review by iulianpascalau — DISMISSED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

File: process/transactions/scrsDataToTransactions.go [hunk 4]
```diff
@@ -48,6 +48,11 @@ func (st *scrsDataToTransactions) attachSCRsToTransactionsAndReturnSCRsWithoutTx
 
 func (st *scrsDataToTransactions) addScResultInfoIntoTx(dbScResult *data.ScResult, tx *data.Transaction) {
 	tx.SmartContractResults = append(tx.SmartContractResults, dbScResult)
+	if isRelayedTx(tx) && len(tx.SmartContractResults) == 1 {
```
Comment (bogdan-rosianu) [inline process/transactions/scrsDataToTransactions.go:51]: can you put this condition into a variable with an appropriate name?

**Review by bogdan-rosianu — DISMISSED** [review]

File: process/transactions/scrsDataToTransactions.go (hunk 4 above)
Comment (miiu96) [reply process/transactions/scrsDataToTransactions.go]: done.

**Review by bogdan-rosianu — APPROVED** [review]

File: mock/dbTransactionsHandlerStub.go [hunk 5]
```diff
@@ -16,6 +16,10 @@ type DBTransactionProcessorStub struct {
 	SerializeScResultsCalled             func(scrs []*data.ScResult) ([]*bytes.Buffer, error)
 }
 
+func (tps *DBTransactionProcessorStub) SerializeTransactionWithRefund(_ map[string]*data.Transaction, _ map[string]*data.RefundData) ([]*bytes.Buffer, error) {
```
Comment (iulianpascalau) [inline mock/dbTransactionsHandlerStub.go:19]: missing comment

**Pull request merged**